From 33311ed71175eedbcd3434265b023dd4ad9054f1 Mon Sep 17 00:00:00 2001
From: Mathis MARION <mathis.marion@grenoble-inp.org>
Date: Wed, 18 Aug 2021 10:49:01 +0200
Subject: [PATCH 2/2] Read access to some Si570 registers

Added read access to HSDIV, N1 and RFREQ (separated in integer and
fractionnal parts).
These can be seen in /sys/bus/zio/adc-100m14b-XXXX/cset0.
---
 kernel/fa-i2c.h              | 16 ++++++++++++++++
 kernel/fa-zio-drv.c          | 23 +++++++++++++++++++++++
 kernel/fmc-adc-100m14b4cha.h |  6 ++++++
 3 files changed, 45 insertions(+)

diff --git a/kernel/fa-i2c.h b/kernel/fa-i2c.h
index 9a0fbf9..ecac066 100644
--- a/kernel/fa-i2c.h
+++ b/kernel/fa-i2c.h
@@ -1,6 +1,22 @@
 #ifndef FA_I2C_H
 #define FA_I2C_H
 
+// Si570 register offsets
+#define SI570_HSDIV_2_0_N1_6_2		7
+#define SI570_N1_1_0_RFREQ_37_32	8
+#define SI570_RFREQ_31_24		9
+#define SI570_RFREQ_23_16		10
+#define SI570_RFREQ_15_8		11
+#define SI570_RFREQ_7_0			12
+#define SI570_7PPM_HSDIV_2_0_N1_6_2	13
+#define SI570_7PPM_N1_1_0_RFREQ_37_32	14
+#define SI570_7PPM_RFREQ_31_24		15
+#define SI570_7PPM_RFREQ_23_16		16
+#define SI570_7PPM_RFREQ_15_8		17
+#define SI570_7PPM_RFREQ_7_0		18
+#define SI570_RST_FRZ_MEMCTR		135
+#define SI570_FREEZE_DCO		137
+
 // I2C master register flags
 #define I2C_MASTER_CTR_EN	(1 << 7)
 
diff --git a/kernel/fa-zio-drv.c b/kernel/fa-zio-drv.c
index ba41928..04d389b 100644
--- a/kernel/fa-zio-drv.c
+++ b/kernel/fa-zio-drv.c
@@ -15,6 +15,7 @@
 #include <asm/byteorder.h>
 
 #include "fmc-adc-100m14b4cha.h"
+#include "fa-i2c.h"
 
 ZIO_PARAM_BUFFER(adc_buffer);
 
@@ -124,6 +125,11 @@ static struct zio_attribute zfad_cset_ext_zattr[] = {
 	ZIO_PARAM_EXT("max-sample-mshot", ZIO_RO_PERM, ZFA_MULT_MAX_SAMP, 0),
 	ZIO_PARAM_EXT("sample-counter", ZIO_RO_PERM, ZFAT_CNT, 0),
 	ZIO_PARAM_EXT("test-data-pattern", ZIO_RW_PERM, ZFAT_ADC_TST_PATTERN, 0),
+
+	ZIO_PARAM_EXT("si570-hsdiv",      ZIO_RO_PERM, ZFA_SI570_HSDIV, 0),
+	ZIO_PARAM_EXT("si570-n1",         ZIO_RO_PERM, ZFA_SI570_N1, 0),
+	ZIO_PARAM_EXT("si570-rfreq-int",  ZIO_RO_PERM, ZFA_SI570_RFREQ_INT, 0),
+	ZIO_PARAM_EXT("si570-rfreq-frac", ZIO_RO_PERM, ZFA_SI570_RFREQ_FRAC, 0),
 };
 
 #if 0 /* FIXME Unused until TLV control will be available */
@@ -407,6 +413,23 @@ static int zfad_info_get(struct device *dev, struct zio_attribute *zattr,
 		i = (int16_t)(*usr_val); /* now signed integer */
 		*usr_val = i;
 		return 0;
+	case ZFA_SI570_HSDIV:
+		*usr_val = fa_i2c_read(fa, SI570_HSDIV_2_0_N1_6_2) >> 5;
+		return 0;
+	case ZFA_SI570_N1:
+		*usr_val = ((fa_i2c_read(fa, SI570_HSDIV_2_0_N1_6_2) & 0b00011111) << 2) |
+			   (fa_i2c_read(fa, SI570_N1_1_0_RFREQ_37_32) >> 6);
+		return 0;
+	case ZFA_SI570_RFREQ_INT:
+		*usr_val = ((fa_i2c_read(fa, SI570_N1_1_0_RFREQ_37_32) & 0b00111111) << 4) |
+			   (fa_i2c_read(fa, SI570_RFREQ_31_24) >> 4);
+		return 0;
+	case ZFA_SI570_RFREQ_FRAC:
+		*usr_val = ((fa_i2c_read(fa, SI570_RFREQ_31_24) & 0b00001111) << 24) |
+			   (fa_i2c_read(fa, SI570_RFREQ_23_16) << 16) |
+			   (fa_i2c_read(fa, SI570_RFREQ_15_8) << 8) |
+			   fa_i2c_read(fa, SI570_RFREQ_7_0);
+		return 0;
 	default:
 		reg_index = zattr->id;
 	}
diff --git a/kernel/fmc-adc-100m14b4cha.h b/kernel/fmc-adc-100m14b4cha.h
index 88ff8ba..891ce34 100644
--- a/kernel/fmc-adc-100m14b4cha.h
+++ b/kernel/fmc-adc-100m14b4cha.h
@@ -371,6 +371,12 @@ enum zfadc_dregs_enum {
 	ZFA_I2C_MASTER_CR,
 	ZFA_I2C_MASTER_SR,
 
+	/* Si570 registers */
+	ZFA_SI570_HSDIV,
+	ZFA_SI570_N1,
+	ZFA_SI570_RFREQ_INT,
+	ZFA_SI570_RFREQ_FRAC,
+
 	ZFA_HW_PARAM_COMMON_LAST,
 };
 
-- 
2.27.0

